Algorithm.Framework/Risk/MaximumSectorExposureRiskManagementModel.cs

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect.Algorithm.Framework.Portfolio;
+using QuantConnect.Data.UniverseSelection;
+
+namespace QuantConnect.Algorithm.Framework.Risk
+{
+    /// <summary>
+    /// Provides an implementation of <see cref="IRiskManagementModel"/> that limits
+    /// the sector exposure to the specified percentage
+    /// </summary>
+    public class MaximumSectorExposureRiskManagementModel : IRiskManagementModel
+    {
+        private readonly decimal _maximumSectorExposure;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MaximumSectorExposureRiskManagementModel"/> class
+        /// </summary>
+        /// <param name="maximumSectorExposure">The maximum exposure for any sector, defaults to 20% sector exposure.</param>
+        public MaximumSectorExposureRiskManagementModel(
+            decimal maximumSectorExposure = 0.20m
+            )
+        {
+            _maximumSectorExposure = Math.Abs(maximumSectorExposure);
+        }
+
+        /// <summary>
+        /// Manages the algorithm's risk at each time step
+        /// </summary>
+        /// <param name="algorithm">The algorithm instance</param>
+        /// <param name="targets">The current portfolio targets to be assessed for risk</param>
+        public IEnumerable<IPortfolioTarget> ManageRisk(QCAlgorithmFramework algorithm, IPortfolioTarget[] targets)
+        {
+            var maximumSectorExposureValue = algorithm.Portfolio.TotalPortfolioValue * _maximumSectorExposure;
+
+            var targetCollection = new PortfolioTargetCollection();
+            targetCollection.AddRange(targets);
+
+            // Group the securities by their sector
+            var groupBySector = algorithm.Securities
+                .Where(x => x.Value.Fundamentals.HasFundamentalData)
+                .GroupBy(x => x.Value.Fundamentals.CompanyReference.IndustryTemplateCode);
+
+            foreach (var securities in groupBySector)
+            {
+                // Compute the sector absolute holdings value
+                // If the construction model has created a target, we consider that
+                // value to calculate the security absolute holding value
+                var sectorAbsoluteHoldingsValue = 0m;
+
+                foreach (var security in securities)
+                {
+                    var absoluteHoldingsValue = security.Value.Holdings.AbsoluteHoldingsValue;
+
+                    IPortfolioTarget target;
+                    if (targetCollection.TryGetValue(security.Value.Symbol, out target))
+                    {
+                        absoluteHoldingsValue = security.Value.Price * Math.Abs(target.Quantity) *
+                            security.Value.SymbolProperties.ContractMultiplier *
```

Comment:
Very smart! I think because of this you actually do need to loop over the securities collection since `OnSecuritiesChanged` will say 'removed' even if you hold stock -- in this way it's like a warning that it's been removed but if you hold stock/open orders we won't remove the subscription.

I've opened https://github.com/QuantConnect/Lean/issues/1945 to add a helper to get securities w/ active data feed subscriptions.

Reply:
Thanks. If the security is removed, `Quantity` and, consequently, `AbsoluteHoldingValue` are zero and it won't be an issue, but looping only through active subscription makes is more efficient.